torch/distributions/utils.py

```diff
@@ -182,6 +182,7 @@ def __init__(self, wrapped):
     def __get__(self, instance, obj_type=None):
         if instance is None:
             return self
-        value = self.wrapped(instance)
+        with torch.enable_grad():
```

Comment:
Meh, it's not the cleanest to have this as part of the `lazy_property` contract 😕 On the other hand moving it outside would make it quite error prone.

Reply:
I agree. But unless we want to have grad modes of two different priorities (user set has highest pri), I can't think of a clean solution.

Reply:
Maybe calculating remembering whether it has grad and re-calculating the thing when it's needed but not available?

Reply:
That won't work. The lazy property makes use of some special Python tricks that cause it to go through `__get__` exactly once, and later accesses to the attr will simply take the thing from the `__dict__` without even asking. They would need to become regular properties, and their contract would get even more complicated.

Reply:
If I understand the logic correctly, the logic is to use `__get__` unless there is an attribute of the same name. If you cache `value` in self and make recalculation and the `setattr` conditional on `is_grad_enabled`, you'd get a property that becomes lazy on the first invocation with grads. I'm not sure whether the overhead of calling `__get__` is that large in the grand scheme of things, but if it is regularly used with `no_grad`, I think the memory saving might be worth the compute.